pkgs/os-specific/darwin/apple-source-releases/default.nix

```
@@ -135,125 +138,135 @@ let
     };
   };
 
-  fetchApple = version: sha256: name: let
+  fetchApple = pname: version: sha256: let
     # When cross-compiling, fetchurl depends on libiconv, resulting
     # in an infinite recursion without this. It's not clear why this
     # worked fine when not cross-compiling
-    fetch = if name == "libiconv"
+    fetch = if pname == "libiconv"
       then stdenv.fetchurlBoot
       else fetchurl;
   in fetch {
-    url = "http://www.opensource.apple.com/tarballs/${name}/${name}-${versions.${version}.${name}}.tar.gz";
+    url = "http://www.opensource.apple.com/tarballs/${pname}/${pname}-${version}.tar.gz";
     inherit sha256;
   };
 
-  appleDerivation_ = name: version: sha256: attrs: stdenv.mkDerivation ({
-    inherit version;
-    name = "${name}-${version}";
+  fetchAppleOnSDK = sdkName: sha256: pname: let
+    version = sdkNames.${sdkName}.${pname};
+  in fetchApple pname version sha256;
+
+  appleDerivation_ = pname: version: sdkName: sha256: attrs: stdenv.mkDerivation ({
+    name = "${pname}-${version}-${sdkName}";
     enableParallelBuilding = true;
     meta = {
       platforms = stdenv.lib.platforms.darwin;
     };
   } // (if attrs ? srcs then {} else {
-    src  = fetchApple version sha256 name;
+    src = fetchApple pname version sha256;
   }) // attrs);
 
-  applePackage = namePath: version: sha256:
-    let
-      name = builtins.elemAt (stdenv.lib.splitString "/" namePath) 0;
-      appleDerivation = appleDerivation_ name version sha256;
-      callPackage = pkgs.newScope (packages // pkgs.darwin // { inherit appleDerivation name version; });
-    in callPackage (./. + "/${namePath}");
+  applePackage = namePath: version: sdkName: sha256: let
+    pname = builtins.head (stdenv.lib.splitString "/" namePath);
+    appleDerivation = appleDerivation_ pname version sdkName sha256;
+    callPackage = pkgs.newScope (packages // pkgs.darwin // { inherit appleDerivation; });
+  in callPackage (./. + "/${namePath}");
+
+  applePackageOnSDK = namePath: sdkName: sha256: let
+    pname = builtins.head (stdenv.lib.splitString "/" namePath);
+    version = sdkNames.${sdkName}.${pname} or "";
```

```suggestion
    version = sdkNames.${sdkName}.${pname};
```
Then just change IOKit to use `applePackage "IOKit" null ...`

---

We can't because all the `IOKit`s' version are not listed on `versions` prior to this PR. They are fine to use `sdkName ` as `version` previously, but not now if we require a specific version. It's OK to use an empty string, I will get them correct versions when updating IOKit later.

---

Okay, it can be `applePackage "IOKit" "" ...`. Then we can get rid of this custom condition?

---

OK, I leave a hack in `versions` so we would need this custom condition anymore.